### ISSUE TITLE
Pure compilation is not that pure

### ISSUE THREAD

**Issue by gabrielhdt** [Deducteam/lambdapi#780]
Static symbols such as `Unif_rule.equiv` are modified by `Pure.compile`, in particular their rules. For instance, using `dune utop` (where the prompt is symbolised by the hash `#`)
```ocaml
# open Core
open Term
open Common
open Timed;;
# Library.set_lib_root None;;
- : unit = ()
# List.length @@ !(Unif_rule.equiv.sym_rules);;
- : int = 0
# ignore @@ Compile.Pure.compile_file "tests/OK/unif_hint.lp";;
Loading "/home/hondet/workspace/deducteam/lambdapi/tests/OK/unif_hint.lp" ...
[...]
- : unit = ()
# List.length @@ !(Unif_rule.equiv.sym_rules);;
- : int = 4
```

**Comment by fblanqui** [comment]
@gabrielhdt #826 does not fully fix the problem in fact. Worst, it is incorrect. And I now understand why there was a non-termination problem in #824.

The problem is that **compilation cannot be both pure and correct** if the file that you compile adds rules on a symbol defined in another module. Assume you have a file M0 where f is declared. Assume that you have another file M1 which requires M0 and adds the rule r1 on M0.f and that r1 is necessary to type other symbols declared in M1. When reading M1, you add r1 in M0.f. Removing r1 afterwards is incorrect since the typing of some other symbol of M1 requires r1.

Assume now that you also have a file M2 which requires M0 but not M1 and adds the rule r2 on M0.f and that r2 is necessary to type other symbols declared in M2. Assume finally that, while both r1 and r2 terminate and are confluent, the combination of r1 and r2 does not terminate or is not confluent. If you compile M1 and do not remove r1 from M0.f then, when you compile M2, you get a non-terminating or non-confluent system.

Conclusion: either you remove added rules in dependencies and should rename pure_compile into unsafe_pure_compile, or you do not remove added rules in dependencies and should remove pure_compile.

To better understand your needs, could you please explain in details why you added this function after all? What is the problem you wanted to solve by adding this function? Maybe there is another way to solve it.

**Comment by gabrielhdt** [comment]
I see, it makes sense. But it is still "correct" to typecheck a file without side effects, it seems, so a function `Pure.check` would make sense, wouldn't it? And anyway, the current implementation of the pure wrappers is correct since it does not revert the signatures. They're not strictly pure, but they allow to bypass administrative resets. And in that sense, I think that the test suite is a true use case of this function: we effectively want to typecheck (not necessarily compile) a bunch of lambdapi files without interference between them.

It seems to me that having both functions would be nice, a true pure typechecker, and a pure-up-to-signatures compiler.

The need for a pure-up-to-signatures compiler is motivated by interoperability: if one wants to make a translation from a system A to a system B, we want to load into memory the encoding of A and the encoding of B to have access to the symbols and write the translation in OCaml.

**Comment by fblanqui** [comment]
For the moment, we should at least rename the function and add some comment.
To guarantee purity, we could also have an option that would make the compilation fail if the file tries to add a rule on an external symbol.
Note however that this imposes some constraints in the scenario that you mention: this forbids A and B to be build in a modular way from other simpler modules. And this is typically what we would like to do with the theory U: A and B would be subsets of U, so share many symbols, and A and B would have slightly different sets of rules. To avoid problems, we would have to assume that A and B share no symbols. You would have A.Prf distinct from B.Prf even if, in your translation, you replace A.Prf by B.Prf.
Then, it could be interesting to have a new command on modules: include. `include M` would inline a copy of M.

**Comment by gabrielhdt** [comment]
I don't see where are the constraints: in the case of compilation we must allow side effects on other signature, and in the case of type checking only, we may forbid *persistent* side effects (that is, side effects may happen while typechecking, but they should be reverted at the end).

By the way, if one wants to go parallel one day, it seems that this will be an issue not to forget.

**Comment by fblanqui** [comment]
"fixed" in #831
